### ISSUE TITLE
Map of lists comes up as possible null with known key

### ISSUE THREAD

**Issue by jpschewe** [typetools/checker-framework#4594]
Given that `teamViolations.containsKey(violationTeam)` is `true` and the type of `teamViolations` doesn't have nullable values, shouldn't it be clear that `vs` cannot be null? 

```java
package net.mtu.eggplant.checker;

import java.util.*;

public class MapOfLists {

  public void testMethod() {
    final Map<Integer, List<Integer>> teamViolations = new HashMap<>();

    final List<Integer> vs;
    final int violationTeam = 10;
    if (teamViolations.containsKey(violationTeam)) {
      vs = teamViolations.get(violationTeam);
    } else {
      vs = new LinkedList<>();
      teamViolations.put(violationTeam, vs);
    }
    vs.add(10);

  }
}
```

Yields this warning:
```
/home/jpschewe/projects/checker-bugs/src/main/java/net/mtu/eggplant/checker/MapOfLists.java:19: error: [dereference.of.nullable] dereference of possibly-null reference vs
    vs.add(10);
    ^
```

**Comment by mernst** [comment]
This is expected behavior.
In your case, the warning is a false positive, but in general code like yours could throw a null pointer exception.

Consider the following code:

```java
import java.util.List;
import java.util.Map;

public class Issue4594 {

  public void testPrimitive(Map<Integer, List<Integer>> teamViolations, int violationTeam) {
    if (teamViolations.containsKey(violationTeam)) {
      final List<Integer> vs = teamViolations.get(violationTeam);
      vs.add(10);
    }
  }

  public void testReference(Map<Integer, List<Integer>> teamViolations, Integer violationTeam) {
    if (teamViolations.containsKey(violationTeam)) {
      final List<Integer> vs = teamViolations.get(violationTeam);
      vs.add(10);
    }
  }
}
```

The Nullness Checker outputs a true positive warning for the first method, and correctly issues no warning for the second method:

```java
Issue4594.java:9: error: [dereference.of.nullable] dereference of possibly-null reference vs
      vs.add(10);
      ^
1 error
```

The reason is boxing.  The first method is translated by javac as

```java
  public void testPrimitive(Map<Integer, List<Integer>> teamViolations, int violationTeam) {
    if (teamViolations.containsKey(Integer.valueOf(violationTeam))) {
      final List<Integer> vs = teamViolations.get(Integer.valueOf(violationTeam));
      vs.add(10);
    }
  }
```

Note the two different calls to `Integer.valueOf` to box the primitive.
In general, `Integer.valueOf` may return a different value on each call.
Furthermore, the map `teamViolations` might be an IdentityHashMap.
In that case, the code can indeed throw a NullPointerException.
You can run this program to see that happen:

```
import java.util.ArrayList;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;

public class Issue4594NPE {

  public static void test(Map<Integer, List<Integer>> teamViolations, int violationTeam) {
    Integer i = violationTeam;
    teamViolations.put(i, new ArrayList<>());
    if (teamViolations.containsKey(i)) {
      final List<Integer> vs = teamViolations.get(violationTeam);
      vs.add(10);
    }
  }

  public static void main(String[] args) {
    test(new IdentityHashMap<>(), 1000);
  }
}
```

It is both more efficient and less error-prone to perform boxing once.  The Nullness Checker is nudging you toward such a design.

In your specific case, there is no possible NullPointerException because the Map happens to not be an IdentityHashMap and because, for small integers, boxing returns the same reference rather than a different one.

**Comment by jpschewe** [comment]
Thank you, that explains the issue. I had not thought about the boxing at all.
